rasa/telemetry.py

```diff
@@ -0,0 +1,554 @@
+    """Use a segment write key to create authentication headers for the segment API.
+
+    Args:
+        write_key: Authentication key for segment.
+
+    Returns:
+        Authentication headers for segment.
+    """
+    return {
+        "Authorization": "Basic {}".format(encode_base64(write_key + ":")),
+        "Content-Type": "application/json",
+    }
+
+
+def segment_request_payload(
+    distinct_id: Text,
+    event_name: Text,
+    properties: Dict[Text, Any],
+    context: Dict[Text, Any],
+) -> Dict[Text, Any]:
+    """Compose a valid payload for the segment API.
+
+    Args:
+        distinct_id: Unique telemetry ID.
+        event_name: Name of the event.
+        properties: Values to report along the event.
+        context: Context information about the event.
+
+    Returns:
+        Valid segment payload.
+    """
+    return {
+        "userId": distinct_id,
+        "event": event_name,
+        "properties": properties,
+        "context": context,
+    }
+
+
+def in_continuous_integration() -> bool:
+    """Returns `True` if currently running inside a continuous integration context."""
+    return any(env in os.environ for env in CI_ENVIRONMENT_TELL)
+
+
+def _is_telemetry_debug_enabled() -> bool:
+    """Check if telemetry debug mode is enabled."""
+    return (
+        os.environ.get(TELEMETRY_DEBUG_ENVIRONMENT_VARIABLE, "false").lower() == "true"
+    )
+
+
+def print_telemetry_event(payload: Dict[Text, Any]) -> None:
+    """Print a telemetry events payload to the commandline.
+
+    Args:
+        payload: payload of the event
+    """
+    print("Telemetry Event:")
+    print(json.dumps(payload, indent=2))
+
+
+async def _send_event(
+    distinct_id: Text,
+    event_name: Text,
+    properties: Dict[Text, Any],
+    context: Dict[Text, Any],
+) -> None:
+    """Report the contents of an event to the /track Segment endpoint.
+    Documentation: https://segment.com/docs/sources/server/http/
+
+    Do not call this function from outside telemetry.py! This function does not
+    check if telemetry is enabled or not.
+
+    Args:
+        distinct_id: Unique telemetry ID.
+        event_name: Name of the event.
+        properties: Values to report along the event.
+        context: Context information about the event.
+    """
+
+    payload = segment_request_payload(distinct_id, event_name, properties, context)
+
+    if _is_telemetry_debug_enabled():
+        print_telemetry_event(payload)
+        return
+
+    write_key = telemetry_write_key()
+    if not write_key:
+        # If TELEMETRY_WRITE_KEY is empty or `None`, telemetry has not been
+        # enabled for this build (e.g. because it is running from source)
+        logger.debug("Skipping request to external service: telemetry key not set.")
+        return
+
+    headers = segment_request_header(write_key)
+
+    async with aiohttp.ClientSession() as session:
+        async with session.post(
```

Comment:
does this also catch these when there is no network connection?

Reply:
in case of no connection, no reporting will be done. the exception will be logged to debug